src/main/java/rx/observables/SyncOnSubscribe.java

```diff
@@ -0,0 +1,499 @@
+ * serialize all calls to the downstream subscriber.
+ *
+ * @param <S>
+ *            the type of the user-define state used in {@link #generateState() generateState(S)} ,
+ *            {@link #next(Object, Subscriber) next(S, Subscriber)}, and
+ *            {@link #onUnsubscribe(Object) onUnsubscribe(S)}.
+ * @param <T>
+ *            the type {@code Subscribers} that will be compatible with {@code this}.
+ */
+@Experimental
+public abstract class SyncOnSubscribe<S, T> implements OnSubscribe<T> {
+    /**
+     * Used to signify when there is unsubscribe behavior
+     */
+    protected Action1<? super S> onUnsubscribe;
+
+    /**
+     * Produces an initial state value. The default implementation returns {@code null}.
+     * 
+     * @return the initial state value that will be received by the {@link #next(Object, Subscriber)
+     *         next(S, Subscriber)} function as it's first argument on it's first execution.
+     */
+    protected S generateState() {
+        return null;
+    }
+
+    /**
+     * SyncOnSubscribe iterates over this function to produce values in a way compatible with
+     * back-pressure. Implementations may not call {@link rx.Subscriber#onNext(Object)
+     * subscriber.onNext(T)} concurrently. You may serialize onNext calls over more than one thread,
+     * but concurrent calls to {@code onNext(t)} will result in a runtime exception. Calling
+     * {@link rx.Subscriber#onCompleted() subscriber.onCompleted()} or
+     * {@link rx.Subscriber#onError(Throwable) subscriber.onError(Throwable)} will result in the
+     * downstream subscriber unsubscribing and executing
+     * {@link SyncOnSubscribe#onUnsubscribe(Object) onUnsubscribe(S)}.
+     * 
+     * @param state
+     *            the value produced by {@link #generateState()} on the first execution or the value
+     *            returned from the previous execution of {@code next(Object, Subscriber)}
+     * @param subscriber
+     *            down stream data subscriber to receive onNext, onCompleted, or onError
+     * @return the state value to be received by the next execution of this function
+     */
+    protected abstract S next(S state, Subscriber<? super T> subscriber);
+
+    /**
+     * Executed after the downstream subscriber is unsubscribed. This method is useful to override
+     * to clean up the SyncOnSubscribe state. Note that this will not be executed concurrently with
+     * {@link #next(Object, Subscriber) next}.
+     * 
+     * @param state
+     *            the state value
+     */
+    private void onUnsubscribe(S state) {
+        if (onUnsubscribe != null)
+            onUnsubscribe.call(state);
+    }
+    
+    /**
+     * Adds an action to execute after the subscription has been unsubscribed from.
+     * 
+     * @param onUnsubscribe action to execute.
+     * @return this
+     */
+    public SyncOnSubscribe<S, T> onUnsubscribe(Action1<S> onUnsubscribe) {
```

Comment:
That is actually implemented via the static method calls (see `SyncOnSubscribe#create...` overloads that do not have an onUnsubscribe). Using this method to set an `onUnsubscribe` is not the primary way to create a `SyncOnSubscribe`. This method provides a way to extend and still take advantage of this fast path. Normal usage would be to use the static methods. I'll see what the team thinks of this method. We might also consider converting this into a builder of an `OnSubscribe`.